docs/proposals/cluster-bootstrap-with-gossip.md

```diff
@@ -0,0 +1,153 @@
+
+    3. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "Gossip",
+        "token": "<token>",
+        "peers": []
+    }
+}
+```
+
+2. If --discovery flag is "out-of-band", require --apiserver-urls and --cacert arguments to also be passed.
+
+    4. In this case, disable the gossip server, and go straight into **bootstrapping** mode.
+
+    5. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "OutOfBand",
+        "apiserverURLs": ["https://foo/"],
+        "cacertFile": "/path/to/cacert.pem"
+    }
+}
+```
+
+In future, there could be other --discovery implementations, such as DNS- or discovery service-based.
+
+The Golang Discovery interface should be:
+
+```
+type Discovery interface {
+    Start()
+    Discover() (apiServerUrls []string, cacert x509.Certificate, err error)
+}
+```
+
+Start should be called every time the kubelet starts up, and it should return immediately, in case the discovery mechanism needs to run a persistent service (like a gossip socket).
+
+### Kubelet awakens
+
+The kubelet waits in **pending** mode for the bootstrap file to show up. When it does, it unmarshals the Discovery object from JSON depending on the "kind" field, and calls `Start()` and then `Discover()` on it.
+
+When it does, it's the Discovery implementation's job to move into **bootstrapping**.
+
+#### Masters
+
+Initializing a new master shouldn't require TLS bootstrap, because the master already has privileged access to the API server. New masters have to add their own address into the API server URL list in the mesh case.
+
+#### Nodes
+
+For the out-of-band implementation it should be pretty obvious how to satisfy that interface with an implementation. The gossip implementation is slightly more clever...
+
+### Gossip implementation
+
+As soon as has Discover called on it, it attempts to form a secure mesh network using the token and the peers, using [a simple gossip protocol library](https://github.com/weaveworks/mesh). For more information on the library, see [this talk](http://infoq.com/presentations/weave-mesh) by the authors.
+
+When that happens, kubelet uses [CRDTs](https://github.com/weaveworks/kubelet-mesh/blob/master/state.go) to support gossiping CA certs and lists of URLs of API servers. As soon as a kubelet learns of both of these pieces of information, it returns from the Discover method, and kubelet proceeds to attempt to perform TLS bootstrap against the API server running on the master.
+
+The Discover() method should spin off a goroutine which continues to listen on the mesh network – at least in the case of the master(s) – so that nodes have something to connect to the gossip network on.
```

Comment:
I don't think we'd be able to use this in GKE, since it would require keeping another port open to the internet (right now only the apiserver is externally exposed).